headless/find.go

```diff
@@ -0,0 +1,221 @@
+	matchTypeRegex = "regexp"
+	matchTypeXPath = "xpath"
+	matchTypeJS    = "js"
+	defaultTimeout = "10s"
+)
+
+var (
+	validMatchTypes = map[string]struct{}{
+		matchTypeJS:    {},
+		matchTypeXPath: {},
+		matchTypeRegex: {},
+	}
+)
+
+type MatchRegExp struct {
+	// Selector is the target element if not specified "*" will be used
+	Selector string `json:"selector"`
+
+	// Value is the target value
+	Value string `json:"value"`
+}
+
+type MatchXPath struct {
+	// Selector is the xpath selector
+	Selector string `json:"selector"`
+}
+
+type MatchJS struct {
+	// Value is the javascript value
+	Value string `json:"value"`
+}
+
+type Match struct {
+	Type        string       `json:"type"`
+	MatchRegExp *MatchRegExp `json:"regexp"`
+	MatchXPath  *MatchXPath  `json:"xpath"`
+	MatchJS     *MatchJS     `json:"js"`
+}
+
+type FindRequest struct {
+	Location string `json:"location"`
+	Timeout  string `json:"timeout"`
+	Match    Match  `json:"match"`
+}
+
+type FindRespond struct {
+	Found   bool   `json:"found"`
+	Message string `json:"message,omitempty"`
+}
+
+func errorResp(c *gin.Context, error_code int, err error) {
+	c.JSON(error_code, FindRespond{
+		Message: err.Error(),
+	})
+}
+
+func validate(c *gin.Context) {
+	var req FindRequest
+	if err := c.Bind(&req); err != nil {
+		errorResp(c, http.StatusBadRequest, xerrors.Errorf("Param error"))
+		return
+	}
+
+	if err := checkValidateRequest(&req); err != nil {
+		errorResp(c, http.StatusBadRequest, err)
+		return
+	}
+
+	launcher := newLauncher(LauncherPath)
+	defer launcher.Cleanup()
+	defer launcher.Kill()
+
+	u, err := launcher.Launch()
+	if err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	browser := rod.New().ControlURL(u)
+	if err := browser.Connect(); err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	defer browser.Close()
+
+	page, err := browser.Page(proto.TargetCreateTarget{URL: req.Location})
+	if err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	timeout := req.Timeout
+	if timeout == "" {
+		timeout = defaultTimeout
+	}
+
+	timeoutDuration, err := time.ParseDuration(timeout)
+	if err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+	router := page.HijackRequests()
+
+	resources := []proto.NetworkResourceType{
+		proto.NetworkResourceTypeFont,
+		proto.NetworkResourceTypeImage,
+		proto.NetworkResourceTypeMedia,
+		proto.NetworkResourceTypeStylesheet,
+		proto.NetworkResourceTypeWebSocket, // we don't need websockets to fetch html
+	}
+
+	router.MustAdd("*", func(ctx *rod.Hijack) {
+		if slice.Contains(resources, ctx.Request.Type()) {
+			ctx.Response.Fail(proto.NetworkErrorReasonBlockedByClient)
+			return
+		}
+
+		ctx.ContinueRequest(&proto.FetchContinueRequest{})
+	})
+
+	go router.Run()
+
+	page = page.Timeout(timeoutDuration)
+	if err := page.WaitLoad(); err != nil {
```

Comment:
Will this `WaitLoad()` handles in-page `XHR` correctly? 
I mean, will it "wait" for enough time to let the page load itself completely?

Reply:
After I read the code of `go-rod`, this `WaitLoad()` waits only for `window.onload`. This will cause a bug when the target page has `XHR`. I'll fix this problem.